src/app/editor/contextualListener.js

```diff
@@ -0,0 +1,93 @@
+'use strict'
+var SourceMappingDecoder = require('ethereum-remix').util.SourceMappingDecoder
+var AstWalker = require('ethereum-remix').util.AstWalker
+
+class ContextualListener {
+  constructor (api, events) {
+    this._api = api
+    this._index = {
+      ReferencedDeclarations: {},
+      Expressions: []
+    }
+    this._events = []
+
+    events.compiler.register('compilationFinished', (success, data, source) => {
+      this._stopWarning()
+      this._index = {
+        ReferencedDeclarations: {},
+        Expressions: []
+      }
+      if (success) {
+        this._buildIndex(data, source)
+      }
+    })
+    this.sourceMappingDecoder = new SourceMappingDecoder()
+    this.astWalker = new AstWalker()
+    setInterval(() => {
+      this._warnExpressions(api.getCursorPosition(), api.getCompilationResult())
+    }, 1000)
+  }
+
+  _warnExpressions (cursorPosition, compilationResult) {
+    if (this.currentPosition === cursorPosition) return
+    this._stopWarning()
```

Comment:
Hm, this should perhaps also be called directly whenever we move the cursor out of the current source location.

Reply:
this is the case